File: iosApp/iosApp/MainViewModel.swift
//
//  MainViewModel.swift
//  iosApp
//
//  Created by Farah Atieh on 22/05/2023.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import KMPNativeCoroutinesRxSwift
import shared
import RxSwift

class MainViewModel: ObservableObject {
    private let repository = MealsRepository.init()
    private let getMeals = GetMealsUseCase.init()
    private let fetchMeals = FetchMealsUseCase.init()
    private let onToggleFavouriteState = ToggleFavouriteStateUseCase.init()
    @Published
    private(set) var state = State.LOADING
    @Published
    var shouldFilterFavourites = false
    @Published
    private(set) var filteredMeals: [Meal] = []
    @Published
    private var meals: [Meal] = []
    private let disposeBag = DisposeBag()

    init(){
        createObservable(for: repository.mealsFlow)
            .subscribe(onNext: { meals in
                DispatchQueue.main.async {
                    self.meals = meals
                }
            }).disposed(by: disposeBag)
        
        
        $meals.combineLatest($shouldFilterFavourites, {
            meals, shouldFilterFavourites -> [Meal] in
            
            var result: [Meal] = []
            
            if(shouldFilterFavourites){
                result.append(contentsOf:
                                meals.filter{$0.isFavourite})
            }else {
                result.append(contentsOf: meals)
            }
            
            if(result.isEmpty){
                self.state = State.EMPTY
            }else {
                self.state = State.NORMAL
            }
            
            return result
        }).assign(to: &$filteredMeals)
    }
    
    func getData(){
        state = State.LOADING
        createSingle(for: getMeals.invoke(parameterValue: getRandomChar()))
            .subscribe(
            onSuccess: { _ in
                DispatchQueue.main.async {
                    self.state = State.NORMAL
                }
            }, onFailure: { error in
                DispatchQueue.main.async {
                    self.state = State.ERROR
                }
            }).disposed(by: disposeBag)
    }
    
    func fetchData(){
        state = State.LOADING
        createSingle(for: fetchMeals.invoke(parameterValue: getRandomChar()))
            .subscribe(onSuccess: { _ in
                DispatchQueue.main.async {
                    self.state = State.NORMAL
                }
            }, onFailure: {error in
                DispatchQueue.main.async {
                    self.state = State.ERROR
                }
            }).disposed(by: disposeBag)
    }
    
    func onFavouriteTapped(meal : Meal){
        createSingle(for: onToggleFavouriteState.invoke(meal: meal))
            .subscribe()
            .disposed(by: disposeBag)
    }
    
    // Use this method to generate a random number between 65 and 90 (ASCII values for A to Z)
    private func getRandomChar() -> String {
        let asciiValue = Int.random(in: 65...90)
        guard let char = UnicodeScalar(asciiValue) else {
            return ""
        }
        return String(char)
    }
}

File: shared/src/commonMain/kotlin/com/example/mymealsapp/viewmodel/MainViewModel.kt
package com.example.mymealsapp.viewmodel

import com.example.mymealsapp.data.Meal
import com.example.mymealsapp.data.usecases.FetchMealsUseCase
import com.example.mymealsapp.data.usecases.GetMealsUseCase
import com.example.mymealsapp.data.usecases.ToggleFavouriteStateUseCase
import com.example.mymealsapp.repository.MealsRepository
import com.example.mymealsapp.shared.ui.CoroutineViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject
import kotlin.random.Random

class MainViewModel : CoroutineViewModel(), KoinComponent {

    private val mealsRepository: MealsRepository by inject()
    private val getMeals: GetMealsUseCase by inject()
    private val fetchMeals: FetchMealsUseCase by inject()
    private val onToggleFavouriteState: ToggleFavouriteStateUseCase by inject()

    private val _state = MutableStateFlow(State.LOADING)
    val state: StateFlow<State> = _state

    private val _isRefreshing = MutableStateFlow(false)
    val isRefreshing: StateFlow<Boolean> = _isRefreshing

    private val _shouldFilterFavourites = MutableStateFlow(false)
    val shouldFilterFavourites: StateFlow<Boolean> = _shouldFilterFavourites

    val meals = mealsRepository.meals
        .combine(_shouldFilterFavourites) { meals, shouldFilterFavourites ->
            if (shouldFilterFavourites) {
                meals.filter { meal ->
                    meal.isFavourite
                }
            } else {
                meals
            }.also {
                _state.value =
                    if (it.isEmpty()) State.EMPTY else State.NORMAL
            }
        }.stateIn(
            coroutineScope,
            SharingStarted.WhileSubscribed(),
            emptyList()
        )

    init {
        loadDate()
    }

    fun loadDate(isForceRefresh: Boolean = false) {
        val parameterValue = getRandomChar()
        val getData: suspend () -> List<Meal> = {
            if (isForceRefresh) {
                fetchMeals.invoke(parameterValue)
            } else {
                getMeals.invoke(parameterValue)
            }
        }

        if (isForceRefresh) {
            _isRefreshing.value = true
        } else {
            _state.value = State.LOADING
        }

        coroutineScope.launch {
            _state.value = try {
                getData()
                State.NORMAL
            } catch (e: Exception) {
                State.ERROR
            }
            _isRefreshing.value = false
        }
    }

    fun refresh() {
        loadDate(true)
    }

    fun onToggleFavouriteFilter() {
        _shouldFilterFavourites.value = ! shouldFilterFavourites.value
    }

    fun onFavouriteTapped(meal: Meal) {
        coroutineScope.launch {
            try {
                onToggleFavouriteState(meal)
            } catch (e: Exception) {
                _state.emit(State.ERROR)
            }
        }
    }

    /**
     * Use this method to generate a random number between 65 and 90 (ASCII values for A to Z)
     */
    private fun getRandomChar(): String {
        val asciiValue = Random.nextInt(65, 91)
        return "${asciiValue.toChar()}"
    }

    @Suppress("unused")
    fun observeMeals(onChange: (List<Meal>) -> Unit){
        meals.onEach {
            onChange(it)
        }.launchIn(coroutineScope)
    }

    @Suppress("unused")
    fun observeState(onChange: (State) -> Unit){
        state.onEach {
            onChange(it)
        }.launchIn(coroutineScope)
    }

    enum class State {
        LOADING,
        NORMAL,
        ERROR,
        EMPTY
    }
}